crates/wysiwyg/src/tests/testutils_composer_model.rs

```diff
@@ -237,55 +236,87 @@ pub fn tx(model: &ComposerModel<u16>) -> String {
     let range = dom.find_range(state.start.into(), state.end.into());
     let is_reversed = state.start > state.end;
 
+    let (first, last): (usize, usize) = if state.start > state.end {
+        (state.end.into(), state.start.into())
+    } else {
+        (state.start.into(), state.end.into())
+    };
+
+    let mut done_last = false;
+
     match range {
         Range::SameNode(range) => {
-            let node = dom.lookup_node_mut(range.node_handle.clone());
-            match node {
+            let mut node = dom.lookup_node_mut(range.node_handle.clone());
+            match &mut node {
                 DomNode::Container(_) => {
                     panic!("Don't know how to tx in a non-text node")
                 }
-                DomNode::Text(text_node) => {
-                    update_text_node_with_cursor(text_node, &range)
-                }
-            }
+                DomNode::Text(n) => update_text_node_with_cursor(n, &range),
+            };
+            done_last = true;
         }
         Range::NoNode => panic!("No node!"),
         Range::MultipleNodes(range) => {
-            // Update start node selection
-            let start_text_node_location = range
-                .locations
-                .iter()
-                .find(|location| location.location_type == Start);
-
-            if let Some(start) = start_text_node_location {
-                if let DomNode::Text(n) =
-                    dom.lookup_node_mut(start.node_handle.clone())
-                {
-                    update_text_nodes_start_with_selection(
-                        n,
-                        &start,
-                        is_reversed,
-                    );
+            let mut i = 0;
+            let mut done_first = false;
+            for loc in range.locations {
+                let mut node = dom.lookup_node_mut(loc.node_handle.clone());
+                match &mut node {
+                    DomNode::Container(_) => {}
+                    DomNode::Text(n) => {
+                        i += n.data().len();
+                        let mut do_start = false;
+                        let mut do_end = false;
+                        if first < i && !done_first {
+                            if is_reversed {
+                                do_end = true;
+                            } else {
+                                do_start = true;
+                            }
+                            done_first = true;
+                        }
+                        if last < i && !done_last {
+                            if is_reversed {
+                                do_start = true;
+                            } else {
+                                do_end = true;
+                            }
+                            done_last = true;
+                        }
```

Comment:
Thank you! I tried to clarify the code and ended up restructuring it quite a lot - let me know what you think.

Reply:
Returning the ending cursor string and applying it if needed is a bit weird, couldn't that be done inside `write_selection_multi`? If it's because it's easier to apply it at the end when it's already a string, maybe both `update_text_node_with_cursor` and that fn could return `dom.string()` + any needed suffix?